### PR TITLE
Support a default tooltip

### PR DESCRIPTION
**Pull request by ajduberstein**
For #3482

### CODE REVIEW

File: modules/core/src/lib/deck.js [hunk 1]
```diff
@@ -559,8 +560,15 @@ export default class Deck {
           _pickRequest
         )
       );
+      const shouldGenerateInfo = _pickRequest.callback || (result && this.props.getTooltip);
```
Comment (Pessimistress) [inline modules/core/src/lib/deck.js]: Under what circumstances is `result` null?

File: modules/core/src/lib/deck.js [hunk 2]
```diff
@@ -559,8 +560,15 @@ export default class Deck {
...
+      let pickedInfo;
+      if (shouldGenerateInfo) {
+        pickedInfo = result.find(info => info.index >= 0) || emptyInfo;
+      }
+      if (result && this.props.getTooltip) {
```
Comment (Pessimistress) [inline modules/core/src/lib/deck.js]: Ditto

File: modules/core/src/lib/tooltip.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+/* global document */
+const defaultStyle = {
+  zIndex: 1001,
+  maxHeight: '500px',
+  position: 'absolute',
+  pointerEvents: 'none',
+  fontSize: '11px',
+  maxWidth: '500px',
+  fontFamily: 'ff-clan-web-pro, "Helvetica Neue", Helvetica, sans-serif',
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Use public fonts please.

File: modules/core/src/lib/tooltip.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    const el = this.el;
+    if (pickedInfo && pickedInfo.picked) {
+      while (el.firstChild) {
+        el.removeChild(el.firstChild);
+      }
+      const displayInfo = processTooltip(pickedInfo.object);
+
+      if (typeof displayInfo === 'string') {
+        el.innerText = displayInfo;
+      } else if (displayInfo === null) {
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Use `!displayInfo` for safety

File: modules/core/src/lib/tooltip.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        el.style.display = 'none';
+      } else {
+        if (displayInfo.text) {
+          el.innerText = displayInfo.text;
+        }
+        if (displayInfo.html) {
+          el.innerHTML = displayInfo.html;
+        }
+        el.className = displayInfo.className || this.el.className;
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: `displayInfo.className` can be `''`

File: modules/core/src/lib/tooltip.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        el.removeChild(el.firstChild);
+      }
+      const displayInfo = processTooltip(pickedInfo.object);
+
+      if (typeof displayInfo === 'string') {
+        el.innerText = displayInfo;
+      } else if (displayInfo === null) {
+        el.style.display = 'none';
+      } else {
+        if (displayInfo.text) {
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: `text` or `html` can be `''`

File: test/modules/core/lib/tooltip.spec.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import Tooltip from '@deck.gl/core/lib/tooltip';
+import test from 'tape-catch';
+
+const {JSDOM} = require('jsdom');
```
Comment (Pessimistress) [inline test/modules/core/lib/tooltip.spec.js]: This is not going to work in the browser.

File: modules/core/src/lib/deck.js [hunk 8]
```diff
@@ -584,6 +592,8 @@ export default class Deck {
       trackContextState(gl, {enable: true, copyState: true});
     }
 
+    this.tooltip = new Tooltip(this.canvas);
```
Comment (Pessimistress) [inline modules/core/src/lib/deck.js]: Tooltip dom element must be removed when deck is finalized.

File: modules/jupyter-widget/src/utils.js [hunk 9]
```diff
@@ -23,9 +23,9 @@ export function updateDeck(inputJSON, {jsonConverter, deckConfig}) {
 }
 
 export function initDeck({mapboxApiKey, container, jsonInput}, onComplete, handleClick) {
-  require(['mapbox-gl', 'h3', 'S2'], mapboxgl => {
-    require(['deck.gl'], deckgl => {
-      try {
+  try {
```
Comment (Pessimistress) [inline modules/jupyter-widget/src/utils.js]: - If this is not related to the tooltip feature, make the change in another PR.
- try... catch does not apply to errors inside the callback.
- Make a comment on what might break.

File: modules/core/src/lib/tooltip.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        el.removeChild(el.firstChild);
+      }
+      const displayInfo = processTooltip(pickedInfo.object);
+
+      if (typeof displayInfo === 'string') {
+        el.innerText = displayInfo;
+      } else if (!displayInfo) {
+        el.style.display = 'none';
+      } else {
+        if (displayInfo.text !== undefined) {
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Use `'text' in displayInfo`

File: modules/core/src/lib/tooltip.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+          el.innerText = displayInfo.text;
+        }
+        if (displayInfo.html !== undefined) {
+          el.innerHTML = displayInfo.html;
+        }
+        el.className =
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Use a if block.

File: modules/core/src/lib/tooltip.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+/* global document */
+const defaultStyle = {
+  zIndex: 1001,
+  maxHeight: '500px',
+  position: 'absolute',
+  pointerEvents: 'none',
+  fontSize: '11px',
+  maxWidth: '500px',
+  fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
+  lineHeight: 1.71429,
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Why this arbitrary number?

File: modules/core/src/lib/tooltip.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  overflowX: 'auto',
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: - What's the point of having a scrollbar if pointer events are disabled?
- Use `text-overflow: ellipsis`? https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow

File: examples/experimental/json-browser/app.js [hunk 14]
```diff
@@ -215,6 +215,10 @@ export default class App extends Component {
                 id="json-deck"
                 {...jsonProps}
                 viewState={viewState}
+                pickingRadius={5}
+                getTooltip={object => {
+                  return {'Num points': object.elevationValue};
```
Comment (Pessimistress) [inline examples/experimental/json-browser/app.js]: This is only going to work with the `HexagonLayer`. Put something more generic here?

File: modules/core/src/lib/deck.js [hunk 15]
```diff
@@ -559,8 +565,15 @@ export default class Deck {
           _pickRequest
         )
       );
+      const shouldGenerateInfo = _pickRequest.callback || this.props.getTooltip;
+      let pickedInfo;
```
Comment (Pessimistress) [inline modules/core/src/lib/deck.js]: Nit: `const pickedInfo = shouldGenerateInfo && result.find(...)`

File: modules/core/src/lib/tooltip.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+  top: 0,
+  left: 0,
+  display: 'none'
+};
+
+export default class Tooltip {
+  constructor(canvas) {
+    const canvasParent = canvas.parentElement;
+
+    if (!this.el && canvasParent) {
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Redundant `this.el` check

File: modules/core/src/lib/tooltip.js [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        if ('text' in displayInfo) {
+          el.innerText = displayInfo.text;
+        }
+        if ('html' in displayInfo) {
+          el.innerHTML = displayInfo.html;
+        }
+        if ('className' in displayInfo) {
+          el.className = displayInfo.className;
+        } else {
+          el.className = this.el.className;
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Setting `className` to itself?

File: modules/core/src/lib/tooltip.js [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        Object.assign(el.style, displayInfo.style);
+      }
+      el.style.display = 'inline-block';
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: `block` is fine

File: examples/experimental/json-browser/app.js (hunk 14 above)
Comment (ajduberstein) [reply examples/experimental/json-browser/app.js]: I think I'll remove the example here in favor of putting the tooltips in our other examples

File: test/modules/core/lib/tooltip.spec.js [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  return {
+    text: `Number of points: ${object.elevationValue}`
+  };
+}
+
+test('Tooltip#constructor', t => {
+  const tooltip = setup(); // eslint-disable-line
+  t.ok(document.getElementsByClassName('tooltip'), 'Tooltip exists in document');
+  t.equals(document.getElementsByClassName('tooltip')[0].style.top, '0px');
+  teardown();
```
Comment (Pessimistress) [inline test/modules/core/lib/tooltip.spec.js]: Test `tooltip.remove()`

File: examples/experimental/json-browser/app.js (hunk 14 above)
Comment (ajduberstein) [reply examples/experimental/json-browser/app.js]: Unless you think it makes sense to roll into this PR, I'll address the examples in a separate one

File: modules/core/src/lib/deck.js [hunk 20]
```diff
@@ -559,8 +565,15 @@ export default class Deck {
...
+      if (shouldGenerateInfo) {
+        pickedInfo = result.find(info => info.index >= 0) || emptyInfo;
+      }
+      if (this.props.getTooltip) {
+        this.tooltip.setTooltip(this.props.getTooltip, pickedInfo);
```
Comment (Pessimistress) [inline modules/core/src/lib/deck.js]: Just call `getTooltip` here - the `Tooltip` class does not have to know the user-facing interface.

File: modules/core/src/lib/tooltip.js [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+      this.el = document.createElement('div');
+      this.el.className = 'tooltip';
+      Object.assign(this.el.style, defaultStyle);
+      canvasParent.appendChild(this.el);
+    }
+  }
+
+  setTooltip(processTooltip, pickedInfo) {
+    const el = this.el;
+    if (pickedInfo && pickedInfo.picked) {
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: - `pickedInfo` is never undefined
- Even if `picked: false` the user might want to display something (e.g. current coordinates under the pointer). This feature should not presume anything.

File: modules/core/src/lib/tooltip.js [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+      this.el = document.createElement('div');
+      this.el.className = 'deck-tooltip';
+      Object.assign(this.el.style, defaultStyle);
+      canvasParent.appendChild(this.el);
+    }
+  }
+
+  setTooltip(displayInfo, x, y) {
+    const el = this.el;
+    while (el.firstChild) {
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Add a comment - why is this needed?

File: modules/core/src/lib/tooltip.js [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/* global document */
+const defaultStyle = {
+  zIndex: 1,
+  position: 'absolute',
+  pointerEvents: 'none',
+  color: '#a0a7b4',
+  textOverflow: 'ellipsis',
```
Comment (Pessimistress) [inline modules/core/src/lib/tooltip.js]: Drop

File: test/modules/core/lib/tooltip.spec.js [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+test('Tooltip#remove', t => {
+  const tooltip = setup();
+  t.equals(document.getElementsByClassName('deck-tooltip').length, 1, 'Tooltip element present');
+  tooltip.remove();
+  t.equals(
+    document.getElementsByClassName('deck-tooltip').length,
+    0,
+    'Tooltip element successfully removed'
+  );
+  t.end();
```
Comment (Pessimistress) [inline test/modules/core/lib/tooltip.spec.js]: Missing `teardown()`?

File: docs/api-reference/deck.md [hunk 25]
```diff
@@ -152,6 +152,31 @@ By default, the deck canvas captures all touch interactions. This prop is useful
 
 Extra pixels around the pointer to include while picking. This is helpful when rendered objects are difficult to target, for example irregularly shaped icons, small moving circles or interaction by touch. Default `0`.
 
+#### `getTooltip` (Function, optional)
+
+Function that takes a hovered-over point and displays a tooltip.
```
Comment (Pessimistress) [inline docs/api-reference/deck.md]: - Link to the picking info documentation
- Add: if the prop is not specified...

File: docs/api-reference/deck.md [hunk 26]
```diff
@@ -152,6 +152,31 @@ By default, the deck canvas captures all touch interactions. This prop is useful
...
+* `text` (String, optional) - Specifies the `innerText` attribute of the tooltip.
+* `html` (String, optional) - Specifies the `innerHTML` attribute of the tooltip. Note that this will override the specified `innerText`.
+* `className` (String, optional) - Class name to attach to the tooltip element.
+* `style` (Object, optional) - An object of CSS styles to apply to the tooltip element, which can override the default styling.
+
+By default, the tooltip has the following CSS style:
+  
+```css
+z-index: 1;
+position: 'absolute';
```
Comment (Pessimistress) [inline docs/api-reference/deck.md]: Drop quote

File: docs/api-reference/deck.md [hunk 27]
```diff
@@ -152,6 +152,31 @@ By default, the deck canvas captures all touch interactions. This prop is useful
...
+color: #a0a7b4;
+background-color: #29323c;
+padding: 10px;
+```
+
+The element also has the default class name of `deck-tooltip`.
```
Comment (Pessimistress) [inline docs/api-reference/deck.md]: Move to the `className` line
